examples/emojisearch/assembly/index.ts

```diff
@@ -0,0 +1,173 @@
+}
+
+export function getEmojiFromString(text: string): string {
+  if (
+    text.length >= 2 &&
+    0xd800 <= text.charCodeAt(0) &&
+    text.charCodeAt(0) <= 0xdbff
+  ) {
+    // The first character is a high surrogate, return the first two characters
+    return text.substring(0, 2);
+  } else {
+    // The first character is not a high surrogate, return the first character
+    return text.substring(0, 1);
+  }
+}
+
+function generateList<TData>(
+  instruction: string,
+  text: string,
+  sample: TData,
+): TData[] {
+  // Prompt trick: ask for a simple JSON object.
+  const modifiedInstruction =
+    "Only respond with valid JSON object in this format:\n" +
+    JSON.stringify(sample) +
+    "\n" +
+    instruction;
+
+  const model = models.getModel<ChatModel>(generationModelName);
+  const input = model.createInput([
+    new SystemMessage(modifiedInstruction),
+    new UserMessage(text),
+  ]);
+
+  input.responseFormat = ResponseFormat.Json;
+
+  const output = model.invoke(input);
+
+  // The output should contain the JSON string we asked for.
+  const json = output.choices[0].message.content.trim();
+
+  return JSON.parse<TData[]>(json);
+}
+
+export function generateEmojiDescriptions(): string[] {
+  let allEmojis: string[] = [];
+  const batchSize: i32 = 10;
+  for (let i: i32 = 0; i < starterEmojis.length; i += batchSize) {
+    const end: i32 = min(i + batchSize, starterEmojis.length);
+    const batch: string[] = starterEmojis.slice(i, end);
+
+    const emojiList = generateList<string>(
+      "write the emoji. It must precisely follow the sample",
+      batch.join(" "),
+      "😭: sobbing face",
+    );
+
+    allEmojis = allEmojis.concat(emojiList);
+  }
+  return allEmojis;
+}
+
+export function upsertAllStarterEmojis(): string {
+  let response = "";
+  const descriptions = generateEmojiDescriptions();
+  for (let i: i32 = 0; i < descriptions.length; i++) {
+    response = insertEmoji(descriptions[i]);
+    if (response !== "success") {
+      return response;
+    }
+  }
+  return response;
+}
+
+function generateText(instruction: string, prompt: string): string {
+  // The imported ChatModel interface follows the OpenAI Chat completion model input format.
+  const model = models.getModel<ChatModel>(generationModelName);
+  const input = model.createInput([
+    new SystemMessage(instruction),
+    new UserMessage(prompt),
+  ]);
+
+  // Here we invoke the model with the input we created.
+  const output = model.invoke(input);
+
+  // The output is also specific to the ChatModel interface.
+  // Here we return the trimmed content of the first choice.
+  return output.choices[0].message.content.trim();
+}
+
+export function insertEmoji(emoji: string): string {
+  // check if emoji already exists
+  const texts = collections.getTexts(emojis);
+  for (let i: i32 = 0; i < texts.keys().length; i++) {
+    const text = texts.get(texts.keys()[i]);
+    const textEmoji = getEmojiFromString(text);
```

Comment:
```suggestion
  const texts = collections.getTexts(emojis).values();
  for (let i: i32 = 0; i < texts.length; i++) {
    const textEmoji = getEmojiFromString(texts[i]);
```

- You weren't ever doing anything with the key
- Calling `.keys()` is expensive (relatively), so you don't want to do it on each pass of the for loop.

Reply:
Also, did you intend to do something with the key? If not, why is this a map?

Reply:
for getTexts in general? so that the user can store the ids with it if they need to. i found it nice to see everything ive stored in the collection at once

Reply:
its the first thing that i wanted while writing the examples so its probably something users would want too